pkg/querier/merge_queryable.go

```diff
@@ -0,0 +1,280 @@
+func (m *mergeQuerier) LabelValues(name string) ([]string, storage.Warnings, error) {
+	if name == string(m.tenantLabelName) {
+		return m.tenantIDs, nil, nil
+	}
+
+	// ensure the name of a retained tenant id label gets handled under the
+	// origignal label name
+	if name == retainExistingPrefix+string(m.tenantLabelName) {
+		name = string(m.tenantLabelName)
+	}
+
+	var funcs = make([]stringSliceFunc, len(m.tenantIDs))
+	for pos := range m.tenantIDs {
+		querier := m.queriers[pos]
+		funcs[pos] = func() ([]string, storage.Warnings, error) {
+			return querier.LabelValues(name)
+		}
+	}
+	return mergeDistinctStringSlice(funcs...)
+}
+
+// LabelNames returns all the unique label names present in the block in sorted order.
+func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
+	var funcs = make([]stringSliceFunc, len(m.tenantIDs))
+	for pos := range m.tenantIDs {
+		querier := m.queriers[pos]
+		funcs[pos] = querier.LabelNames
+	}
+	labelNames, warnings, err := mergeDistinctStringSlice(funcs...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// check if the tenant label exists in the original result
+	var tenantLabelExists bool
+	for _, labelName := range labelNames {
+		if labelName == string(m.tenantLabelName) {
+			tenantLabelExists = true
+			break
+		}
+	}
+
+	if tenantLabelExists {
+		// the tenant label exists in the original result, so it had been retained
+		labelNames = append(labelNames, retainExistingPrefix+string(m.tenantLabelName))
+	} else {
+		// the tenant label was not existing, so it need to get added
+		labelNames = append(labelNames, string(m.tenantLabelName))
+	}
+	sort.Strings(labelNames)
+
+	return labelNames, warnings, nil
+
+}
+
+// Close releases the resources of the Querier.
+func (m *mergeQuerier) Close() error {
+	var errs prometheus.MultiError
+	for pos := range m.tenantIDs {
+		if err := m.queriers[pos].Close(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to close querier for %s: %w", m.tenantIDs, err))
+		}
+	}
+	return errs.MaybeUnwrap()
+}
+
+// Select returns a set of series that matches the given label matchers.
+// Caller can specify if it requires returned series to be sorted. Prefer not requiring sorting for better performance.
+// It allows passing hints that can help in optimising select, but it's up to implementation how this is used if used at all.
+func (m *mergeQuerier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+	tenantIDsPos, filteredMatchers := filterValuesByMatchers(string(m.tenantLabelName), m.tenantIDs, matchers...)
+	var seriesSets = make([]storage.SeriesSet, len(tenantIDsPos))
+	for pos, posTenant := range tenantIDsPos {
+		tenantID := m.tenantIDs[posTenant]
+		seriesSets[pos] = &addLabelsSeriesSet{
+			upstream: m.queriers[posTenant].Select(sortSeries, hints, filteredMatchers...),
+			labels: labels.Labels{
+				{
+					Name:  string(m.tenantLabelName),
+					Value: tenantID,
+				},
+			},
+		}
```

Comment:
**suggestion(optional,if-minor):** have you considered running the `Select` for each tenant concurrently? It's probably not something that is needed for an initial implementation but it may be useful to help speed things up.

Reply:
I have added two TODO regarding concurrency. I would leave it out of the initial implementation if that is ok, but I think it is something that can follow very soon